src/cumulative_imputation_links.py

```diff
@@ -46,7 +46,12 @@ def get_cumulative_links(
 
     dataframe["imputation_group"] = (
         (
-            (dataframe["missing_value"].diff(time_difference) != 0)
+            (
+                dataframe["imputation_marker"]
+                .ne(dataframe["imputation_marker"].shift().bfill())
+                .astype(int)
```

Comment:
Lines 47 to 57 are a bit hard to follow. Consider defining this condition first before applying it

Reply:
You are absolutely right, i defined them in the top, also added a TODO line since similar conditions are applied 2-3 times in other functions